base/src/main/java/proguard/optimize/inline/LambdaImplementationVisitor.java

```diff
@@ -0,0 +1,66 @@
+package proguard.optimize.inline;
+
+import proguard.classfile.*;
+import proguard.classfile.constant.ClassConstant;
+import proguard.classfile.constant.FieldrefConstant;
+import proguard.classfile.constant.visitor.ConstantVisitor;
+import proguard.classfile.visitor.AllMethodVisitor;
+import proguard.classfile.visitor.MemberVisitor;
+
+public class LambdaImplementationVisitor implements ConstantVisitor, MemberVisitor {
+    private final InvokeMethodVisitor invokeMethodVisitor;
+    private Clazz lambdaImplementationClass;
+    public LambdaImplementationVisitor(InvokeMethodVisitor invokeMethodHandler) {
+        this.invokeMethodVisitor = invokeMethodHandler;
+    }
+
+    @Override
+    public void visitFieldrefConstant(Clazz clazz, FieldrefConstant fieldrefConstant) {
+        lambdaImplementationClass = fieldrefConstant.referencedClass;
+        fieldrefConstant.referencedClass.interfaceConstantsAccept(this);
+    }
+
+    @Override
+    public void visitClassConstant(Clazz clazz, ClassConstant referencedClassConstant) {
+        referencedClassConstant.referencedClass.methodsAccept(this);
+    }
+
+    @Override
+    public void visitProgramMethod(ProgramClass interfaceClass, ProgramMethod programMethod)
+    {
+        getLambdaImplementation(interfaceClass, programMethod);
+    }
+
+    @Override
+    public void visitLibraryMethod(LibraryClass interfaceClass, LibraryMethod libraryMethod) {
+        getLambdaImplementation(interfaceClass, libraryMethod);
+    }
+
+    private void getLambdaImplementation(Clazz interfaceClazz, Method method) {
+        String descriptor = method.getDescriptor(interfaceClazz);
+        lambdaImplementationClass.methodAccept("invoke", descriptor, new MemberVisitor() {
+            @Override
+            public void visitProgramMethod(ProgramClass programClass, ProgramMethod programMethod) {
```

Comment:
I'm not entirely following what you are trying to achieve here. Would it not be possible to instead of in the `visitProgramMethod` calling the `getLambdaImplementation` just have this implementation in the `visitProgramMethod`? Also calling a MemberVisitor inside of a MemberVisitor doesn't seem to make sense to me and I don't see how it influences the code execution, this can all be inside of one `visitProgramMethod` right?